test/e2e/proxier.go

```diff
@@ -0,0 +1,176 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/kubernetes/pkg/proxy/iptables"
+	"k8s.io/kubernetes/pkg/util/exec"
+	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	iptablestest "k8s.io/kubernetes/pkg/util/iptables/testing"
+)
+
+func makeNSN(namespace, name string) types.NamespacedName {
+	return types.NamespacedName{Namespace: namespace, Name: name}
+}
+
+type fakeSysctl struct{}
+
+func (f *fakeSysctl) GetSysctl(sysctl string) (int, error) {
+	return 1, nil
+}
+func (f *fakeSysctl) SetSysctl(sysctl string, newVal int) error {
+	return nil
+}
+
+type fakeEventRecorder struct{}
+
+func (f *fakeEventRecorder) Event(object pkgruntime.Object, eventtype, reason, message string) {}
+func (f *fakeEventRecorder) Eventf(object pkgruntime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+func (f *fakeEventRecorder) PastEventf(object pkgruntime.Object, timestamp metav1.Time, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+
+type fakeHealthzUpdater struct{}
+
+func (f *fakeHealthzUpdater) UpdateTimestamp() {}
+
+func NewFakeProxier(ipt utiliptables.Interface) *iptables.Proxier {
+	p, err := iptables.NewProxier(ipt,
+		&fakeSysctl{},
+		&exec.FakeExec{},
+		time.Microsecond,
+		time.Microsecond,
+		false,
+		0,
+		"10.0.0.0/24",
+		"testHostname",
+		net.ParseIP("127.0.0.1"),
+		&fakeEventRecorder{},
+		&fakeHealthzUpdater{},
+	)
+
+	if err != nil {
+		panic("can't create proxy!")
+	}
+	return p
+}
+
+func makeEndpointsMap(proxier *iptables.Proxier, allEndpoints ...*api.Endpoints) {
+	for i := range allEndpoints {
+		proxier.OnEndpointsAdd(allEndpoints[i])
+	}
+
+	proxier.OnEndpointsSynced()
+}
+
+func makeTestService(namespace, name string, svcFunc func(*api.Service)) *api.Service {
+	svc := &api.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec:   api.ServiceSpec{},
+		Status: api.ServiceStatus{},
+	}
+	svcFunc(svc)
+	return svc
+}
+
+func makeServiceMap(proxier *iptables.Proxier, allServices ...*api.Service) {
+	for i := range allServices {
+		proxier.OnServiceAdd(allServices[i])
+	}
+
+	proxier.OnServiceSynced()
+}
+
+var _ = framework.KubeDescribe("Network", func() {
+
+	fr := framework.NewDefaultFramework("network")
+
+	It("should sync valid iptables rules", func() {
+		nodes := framework.GetReadySchedulableNodesOrDie(fr.ClientSet)
+
+		if len(nodes.Items) < 1 {
+			framework.Skipf(
+				"Test requires >= 1 Ready nodes, but there are only %v nodes",
+				len(nodes.Items))
+		}
+
+		selectNode := &nodes.Items[0]
+
+		ipt := iptablestest.NewFake()
+		fp := NewFakeProxier(ipt)
+
+		svcPortName := proxy.ServicePortName{
+			NamespacedName: makeNSN("ns1", "svc1"),
+			Port:           "p80",
+		}
+
+		svcIP := "10.20.30.41"
+		svcPort := 80
+		svcExternalIPs := "50.60.70.81"
+
+		makeServiceMap(fp,
+			makeTestService(svcPortName.Namespace, svcPortName.Name, func(svc *api.Service) {
```

Comment:
Might be nice to see the other Service types (LoadBalancer, NodePort, etc) also get tested.  I don't think that would be a bunch of work, but I could be wrong.